lib/Analysis/GateAnalysis.cc

```diff
@@ -171,24 +170,19 @@ void GateAnalysisImpl::processPhi(PHINode *PN, Instruction *insertionPt) {
   DenseMap<BasicBlock *, Value *> incomingBlockToValue =
       processIncomingValues(PN, insertionPt);
 
+  // Make sure CD blocks are sorted in reverse-topological order. We need this
+  // because we want to process them in order opposite to execution order.
+  auto bb_to_compare = [this](BasicBlock *first, BasicBlock *second) {
+    return m_CDA.getBBTopoIdx(first) > m_CDA.getBBTopoIdx(second);
+  };
   // Collect all blocks the incoming blocks are control dependent on.
-  std::vector<BasicBlock *> cdInfo;
+  std::set<BasicBlock *, decltype(bb_to_compare)> cdInfo(bb_to_compare);
```

Comment:
If I interpret this code correctly, `std::set` is used to ensure that collected basic blocks are sorted.
It is more efficient and simpler to collect basic blocks first into a vector and then sort them. As the previous code was doing.

Reply:
Yes, this is what it's doing the code

Reply:
Not really. The examples I ran were slower with using vector, sorting and then deleting redundant entries.

Reply:
Hmmm... this does not make sense. From complexity perspective, both algorithms are identical. Using a properly sized vector and sorting requires fewer memory allocations. I would suggest profiling the code to see where the time is going.

http://lafstern.org/matt/col1.pdf
see also `boost::flat_set`

Reply:
So for the examples I ran, the vector initially holds around 700,000 entries (`m`) and de-duplication reduces the number of entries to around 4000 (`n`). With a vector the expected complexity is now `m log(m)` and with a set it is `m log(n)`.

Reply:
`flat_set` would be a terrible choice here since the insertions are not in order.

Reply:
Ahh, just saw the insertion time complexity.

Reply:
I am surprised at the number of blocks. If it is that high, you should mark the blocks that are entered in the set and skip adding them a second time. In which case using `set` is more justified. At least a comment to that effect would be welcome so that we can evaluate this in the future.

Reply:
> you should mark the blocks that are entered in the set and skip adding them a second time

I can do this but can you explain how does this helps? Checking if the element is already present should have the same complexity as inserting an already existing element. Maybe the confusion is between a `set` and `multiset`.

Reply:
Also, maybe you would prefer this more: Maintain a hash set to avoid redundant entries and then convert it into a vector and sort it in reverse topological order.

Reply:
since basic blocks have identifiers, I would have used a bit vector. But at that point, your use of `std::set` is justified.

Reply:
Since both pointers and topo indices are unique, yo can deduplicate the values first with a `DenseSet` and then copy it over to a vector and sort. But this shouldn't matter much.

Reply:
Added a comment to reflect the use case of set.